### PR TITLE
Use object ID consistently

### PR DESCRIPTION
**Pull request by timja**
Relates to https://github.com/jenkinsci/azure-ad-plugin/issues/190

Fixes https://github.com/jenkinsci/azure-ad-plugin/issues/278

This seems to fix inconsistencies around auth and uses the approach recommended by Microsoft for identifiers:
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims

i.e never use UPNs for IDs.

Unfortunately this means that users display names will not show up in the jcasc export.
Groups will continue to have their `name (object id)` format though

TODO:
- Port tests from matrix auth
- Check existing JCasC config works
- Review docs
- Check role strategy is okay

### CODE REVIEW

File: src/main/java/com/microsoft/jenkins/azuread/AzureSecurityRealm.java [hunk 1]
```diff
@@ -797,8 +798,7 @@ private String generateDescription(Authentication auth) {
                     + "\nUnique Principal Name: " + user.getUniqueName()
                     + "\nEmail: " + user.getEmail()
                     + "\nObject ID: " + user.getObjectID()
-                    + "\nTenant ID: " + user.getTenantID()
-                    + "\nGroups: " + user.getGroupOIDs() + "\n";
+                    + "\nTenant ID: " + user.getTenantID();
```
[the remark this answers is not in the document]
Comment (timja) [reply src/main/java/com/microsoft/jenkins/azuread/AzureSecurityRealm.java:801]: I'll revert this, the `whoAmI` page also displayed it, so I removed as it seemed unneeded, but I see the user profile page doesn't display it.

**Comment by andrewlorien** [comment]
This does not solve https://github.com/jenkinsci/azure-ad-plugin/issues/253 (and i cannot reproduce https://github.com/jenkinsci/azure-ad-plugin/issues/190)

**Comment by timja** [comment]
> This does not solve #253 (and i cannot reproduce #190)

Using a guest user for me was how I reproduced #190.

**Comment by meiswjn** [comment]
Thanks for the fix!

We have noticed two bugs, however, one of them breaking:
- Favorites have been deleted.
- In Auth Matrix, users that were added before the update can now be added again - but changing anything in their config has no effect. The old user must be configured.

**Comment by timja** [comment]
I've updated the release notes

**Comment by ivy-cst** [comment]
Not so nice, that now all users are duplicated! For our use cases it was nicer to have the e-mail as user id instead of the object id.
I understand that it is technically better to use the object id instead of the UPN.
But in jenkins it ends up that the user sees UPN's and also the are in the urls.
Would it be an option to make this configurable?
For now I have downgraded the plugin.

**Comment by timja** [comment]
Prettiness in URLs won't take precedence over bugs caused by it.

https://github.com/jenkinsci/azure-ad-plugin/issues/392
https://github.com/jenkinsci/azure-ad-plugin/issues/278
and some of https://github.com/jenkinsci/azure-ad-plugin/issues/190

**Comment by timja** [comment]
sure warning coming in https://github.com/jenkinsci/azure-ad-plugin/pull/396

I don't think it will be possible to auto migrate, a script should be possible though if you write one I'm happy to link in release notes

**Comment by andysworkshop** [comment]
This has broken usage of the Jenkins API where previously `user@example.com:$TOKEN` was working for basic auth. Now you have to use object ID in place of email address for auth. We're going to communicate this to our users and patch our UIs that prompt for 'email address' but would appreciate it if the UPN could again be used for API call auth. IMO GUIDs are an internal ugliness that users should not see.

**Comment by timja** [comment]
it might be possible to have it as an option but there are issues with using it as the UPN, feel free to file an issue / PR for it.

**Comment by Louai-Abdelsalam** [comment]
Can the documentation on the plugin's page (specifically this [line](https://plugins.jenkins.io/azure-ad/#:~:text=user%20or%20group.-,This,-plugin%20extends%20the) and this [line](https://plugins.jenkins.io/azure-ad/#:~:text=based%20Authorization%20Strategy-,Just,-keep%20in%20mind)) be updated to reflect display names no longer showing up? my plugin got updated by someone else and I didn't know it, so I spent quite a bit of time super confused as to what happened, and having had the mentioned parts of the documentation updated would've helped me for sure.

**Pull request merged**